### PR TITLE
Add new Hyperlink API+ code refactoring

### PR DESCRIPTION
**Pull request by Takoooooo**
This pull request adds new Api to have clickable links in your MessageBox.To use it you will need to call MessageBox.Avalonia.MessageBoxManager.GetMessageBoxHyperlinkWindow and to initialize HyperlinkContentProvider.If you will add only Alias it would be displayed just in text format without links(as in usual messagebox).If you will initialize also Alias and Url then Alias would be used as hyperlink(clickable).If you will initialize only Hyperlink it would be displayed(clickable).Made also a bit of refactoring but i think it needs to be refactored even more.

### CODE REVIEW

**Comment by Takoooooo** [comment]
appended small fix of content but at the moment i stack a bit on making text selectable

File: src/MessageBox.Avalonia/Extensions/OpenUrl.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            if (!Uri.TryCreate(url, UriKind.Absolute, out var tmp)) return false;
+            return tmp.Scheme == Uri.UriSchemeHttp || tmp.Scheme == Uri.UriSchemeHttps;
+        }
+        public static void OpenUrl(this string url)
+        {
+            if (IsValidUrl(url))
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    //https://stackoverflow.com/a/2796367/241446
+                    using (Process proc = new Process { StartInfo = { UseShellExecute = true, FileName = url } })
```
Comment (rstm-sf) [inline src/MessageBox.Avalonia/Extensions/OpenUrl.cs:22]: It seems that it is better to call the factory method here, which, depending on the OS, will return us an instance

File: src/MessageBox.Avalonia/Extensions/SetStyle.cs [hunk 2]
```diff
@@ -13,46 +13,46 @@ internal static void SetStyle(this Window window, Style style)
             switch (style)
             {
                 case Style.Windows:
-                {
-                    styles.Add(new StyleInclude(
-                            new Uri("avares://MessageBox.Avalonia/Styles/Windows/Windows.xaml"))
-                        {Source = new Uri("avares://MessageBox.Avalonia/Styles/Windows/Windows.xaml")});
-                    break;
-                }
+                    {
```
Comment (rstm-sf) [inline src/MessageBox.Avalonia/Extensions/SetStyle.cs:16]: More like an unnecessary change to the codestyle. These changes are best avoided as they are individual and increase cognitive load. Besides, it seems that it worse :) Please remove such changes

File: src/MessageBox.Avalonia/MessageBoxManager.cs [hunk 3]
```diff
@@ -19,32 +18,41 @@ public static IMsBoxWindow<string> GetMessageBoxCustomWindow(MessageBoxCustomPar
...
-
         public static IMsBoxWindow<ButtonResult> GetMessageBoxStandardWindow(string title, string text,
             ButtonEnum @enum = ButtonEnum.Ok, Icon icon = Icon.None,
             WindowStartupLocation windowStartupLocation = WindowStartupLocation.CenterScreen,
             Style style = Style.None) => GetMessageBoxStandardWindow(new MessageBoxStandardParams
-        {
-            ContentTitle = title, ContentMessage = text, ButtonDefinitions = @enum, Icon = icon, Style = style,
-            WindowStartupLocation = windowStartupLocation
-        });
+            {
```
Comment (rstm-sf) [inline src/MessageBox.Avalonia/MessageBoxManager.cs:41]: This going too far in indent ;)

File: src/MessageBox.Avalonia/Controls/Hyperlink.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+using MessageBox.Avalonia.Extensions;
+
+namespace MessageBox.Avalonia.Controls
+{
+    public class Hyperlink : TextBlock
+    {
+        private string _url;
+        public static readonly DirectProperty<Hyperlink, string> UrlProperty
+            = AvaloniaProperty.RegisterDirect<Hyperlink, string>(nameof(Url), o => o.Url, (o, v) => o.Url = v);
+        public Hyperlink()
```
Comment (rstm-sf) [inline src/MessageBox.Avalonia/Controls/Hyperlink.cs:13]: It's unessential to have the empty constructor

File: src/MessageBox.Avalonia.Example/MainWindow.xaml.cs [hunk 5]
```diff
@@ -14,20 +12,44 @@ public class MainWindow : Window
...
+                WindowStartupLocation = WindowStartupLocation.CenterScreen,
+                ButtonDefinitions = MessageBoxAvaloniaEnums.ButtonEnum.Ok
+            });
+            var messageBoxCustomWindow2 = MessageBox.Avalonia.MessageBoxManager.GetMessageBoxInputWindow(new MessageBoxInputParams()
+            {
+                CanResize = true,
                 WindowStartupLocation = WindowStartupLocation.CenterScreen
-               
             });
+            //var messageBoxCustomWindow2 = MessageBox.Avalonia.MessageBoxManager.GetMessageBoxStandardWindow(new MessageBoxStandardParams()
```
Comment (rstm-sf) [inline src/MessageBox.Avalonia.Example/MainWindow.xaml.cs:44]: Why do we need to have the code commented out?

File: src/MessageBox.Avalonia/BaseWindows/Base/IMsBoxWindow.cs [hunk 6]
```diff
@@ -1,7 +1,7 @@
-using System.Threading.Tasks;
 using Avalonia.Controls;
+using System.Threading.Tasks;
 
-namespace MessageBox.Avalonia.BaseWindows
+namespace MessageBox.Avalonia.BaseWindows.Base
```
Comment (rstm-sf) [inline src/MessageBox.Avalonia/BaseWindows/Base/IMsBoxWindow.cs:4]: This is breaking changes

File: src/MessageBox.Avalonia/BaseWindows/MsBoxCustomWindow.cs [hunk 7]
```diff
@@ -1,32 +0,0 @@
-using System.Threading.Tasks;
-using Avalonia.Controls;
-using MsWindow = MessageBox.Avalonia.Views.MsBoxCustomWindow;
-
-namespace MessageBox.Avalonia.BaseWindows
-{
-    public class MsBoxCustomWindow : IMsBoxWindow<string>
```
Comment (rstm-sf) [inline src/MessageBox.Avalonia/BaseWindows/MsBoxCustomWindow.cs:7]: By deleting / changing the public API, we thereby call breaking changes

File: src/MessageBox.Avalonia/ViewModels/Base/AbstractMsBoxViewModel.cs [hunk 8]
```diff
@@ -36,13 +33,12 @@ public AbstractMsBoxViewModel(AbstractMessageBoxParams @params)
...
             ContentHeader = @params.ContentHeader;
-            ContentMessage = @params.ContentMessage;
             LocationOfMyWindow = @params.WindowStartupLocation;
         }
 
-        public async Task Copy()
-        {
-            await AvaloniaLocator.Current.GetService<IClipboard>().SetTextAsync(ContentMessage);
-        }
+        //public async Task Copy()
```
Comment (rstm-sf) [inline src/MessageBox.Avalonia/ViewModels/Base/AbstractMsBoxViewModel.cs:39]: Why is this commented out?

**Review by rstm-sf — CHANGES_REQUESTED** [review]
Hello!

I did a little code review. In my opinion, there are too many breaking changes in this PR. Can we reduce their number?

File: src/MessageBox.Avalonia.Example/MainWindow.xaml.cs (hunk 5 above)
Comment (Takoooooo) [reply src/MessageBox.Avalonia.Example/MainWindow.xaml.cs:44]: @rstm-sf MainWindow it's only for testing so I don't think wee need to keep code clean here. But if it's needed I may do a cleanup

File: src/MessageBox.Avalonia/BaseWindows/MsBoxCustomWindow.cs (hunk 7 above)
Comment (Takoooooo) [reply src/MessageBox.Avalonia/BaseWindows/MsBoxCustomWindow.cs:7]: @rstm-sf I don't think what it will cause breaking changes.It's just for internal use, User don't call it directly so I may change it easily

File: src/MessageBox.Avalonia/ViewModels/Base/AbstractMsBoxViewModel.cs (hunk 8 above)
Comment (Takoooooo) [reply src/MessageBox.Avalonia/ViewModels/Base/AbstractMsBoxViewModel.cs:39]: @rstm-sf Because there are no references pointing to this method so I decided to comment this out, but not to delete because the developer may need this

File: src/MessageBox.Avalonia/MessageBoxManager.cs (hunk 3 above)
Comment (Takoooooo) [reply src/MessageBox.Avalonia/MessageBoxManager.cs:41]: @rstm-sf lol previously you said what its individual

File: src/MessageBox.Avalonia/BaseWindows/MsBoxCustomWindow.cs (hunk 7 above)
Comment (Takoooooo) [reply src/MessageBox.Avalonia/BaseWindows/MsBoxCustomWindow.cs:7]: i think here we just need to change the access modifier of this API because it doesn't need to be used by users

File: src/MessageBox.Avalonia/BaseWindows/Base/IMsBoxWindow.cs (hunk 6 above)
Comment (Takoooooo) [reply src/MessageBox.Avalonia/BaseWindows/Base/IMsBoxWindow.cs:4]: Here I think we need to converse with repository owner. If our point to provide API with a minimum count of breaking changes we can roll this back

File: src/MessageBox.Avalonia/Extensions/OpenUrl.cs (hunk 1 above)
Comment (Takoooooo) [reply src/MessageBox.Avalonia/Extensions/OpenUrl.cs:22]: How do you want to implement this? May you provide an example?

**Comment by Takoooooo** [comment]
It seems to be impossible to make hyperlink message box content selectable. There is probably a bug.
https://github.com/AvaloniaUI/Avalonia/issues/4571

File: src/MessageBox.Avalonia/Extensions/OpenUrl.cs (hunk 1 above)
Comment (rstm-sf) [reply src/MessageBox.Avalonia/Extensions/OpenUrl.cs:22]:
```cs
public static Process CreateProcess(string url)
{
    const string net46AndOlder = "4.0.30319.42000";
    if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
    {
        return new Process { StartInfo = { UseShellExecute = true, FileName = url } };
    }
    else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
    {
        return new Process { StartInfo = { FileName = "x-www-browser", Arguments = url } };
    }
    else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
    {
        return new Process { StartInfo = { FileName = "open", Arguments = url } };
    }
    else if (net46AndOlder.Equals(Environment.Version.ToString(), StringComparison.InvariantCulture))
    {
        return new Process { StartInfo = { UseShellExecute = true, FileName = url } };    
    }

    throw new NotSupportedException(
        "Opening URL is only available on the following operating systems: Windows, Linux, OSX");
}
```

File: src/MessageBox.Avalonia/MessageBoxManager.cs (hunk 3 above)
Comment (rstm-sf) [reply src/MessageBox.Avalonia/MessageBoxManager.cs:41]: That's right, you walked away from the canon

File: src/MessageBox.Avalonia/ViewModels/Base/AbstractMsBoxViewModel.cs (hunk 8 above)
Comment (rstm-sf) [reply src/MessageBox.Avalonia/ViewModels/Base/AbstractMsBoxViewModel.cs:39]: This is a public method. It is not necessary that it is used in this solution

File: src/MessageBox.Avalonia.Example/MainWindow.xaml.cs (hunk 5 above)
Comment (rstm-sf) [reply src/MessageBox.Avalonia.Example/MainWindow.xaml.cs:44]: This is a sample, so leaving islands of incomprehensible lines is not worth it

File: src/MessageBox.Avalonia/BaseWindows/Base/IMsBoxWindow.cs (hunk 6 above)
Comment (rstm-sf) [reply src/MessageBox.Avalonia/BaseWindows/Base/IMsBoxWindow.cs:4]: Yes, that's right. If this happens, then it needs to be mentioned somewhere to facilitate the transition for library users
